[WGSL] Implement

djg opened the pull request:
#### 7deca01aab70ed11a8fb98b3b4aa48138571cf42
<pre>
[WGSL] Implement parsing for array types and callable array constructors.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244792">https://bugs.webkit.org/show_bug.cgi?id=244792</a>
rdar://99557526

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/TypeDecl.h:
(WGSL::AST::TypeDecl::isArray const):
Add kind and predicate for array types.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseTypeDecl):
Begin parsing of an array type declaration. Defers to parseArrayType to do
actual parsing.
(WGSL::Parser&lt;Lexer&gt;::parseArrayType):
Parse an array type consisting of &apos;array&apos; (&apos;&lt;&apos; TypeDecl (&apos;,&apos;
AdditionExpression)? &apos;&gt;&apos;)?
This is different to the WGSL spec as this recursive descent parser can&apos;t handle
the grammar factored as AdditiveExpression | BitwiseExpression. Making the
element TypeDecl optional allows the code to parse array in
CallableExpression. Validity of array TypeDecls that require element TypeDecl
will be handled when type-checking the AST.
(WGSL::Parser&lt;Lexer&gt;::parsePrimaryExpression):
Accept array type as a CallableExpression.
* Source/WebGPU/WGSL/ParserPrivate.h:
Add prototype for parsing array type
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
- Test the parsing of a WebGPU example triangle vertex &amp; fragment shader that
contains ArrayType, ArrayAccess and CallableExpression of array.
- Add some macros to help alleviate the repeative AST matching patterns.
</pre>
----------------------------------------------------------------------
#### 69da8f7f7d17eecbb909abb58a1b0b4a642b3792
<pre>
[WGSL] Implement lexing for &apos;array&apos; keyword
<a href="https://bugs.webkit.org/show_bug.cgi?id=244791">https://bugs.webkit.org/show_bug.cgi?id=244791</a>
rdar://99557428

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/Lexer.cpp:
(WGSL::Lexer&lt;T&gt;::lex):
Lex &quot;array&quot; -&gt; TokenType::KeywordArray
* Source/WebGPU/WGSL/Token.cpp:
(WGSL::toString):
TokenType::KeywordArray -&gt; &quot;array&quot;
* Source/WebGPU/WGSL/Token.h:
Add token for &apos;array&apos; keyword
* Source/WebGPU/WGSLUnitTests/WGSLLexerTests.mm:
- Update for &quot;array&quot; -&gt; ArrayKeyword
- Add tests for lexing all recognized keywords
</pre>
----------------------------------------------------------------------
#### b1f4ab2f8e7e5a58d815cd6c208eca8698915981
<pre>
[WGSL] Implement parsing for ArrayAccess postfix expression.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244789">https://bugs.webkit.org/show_bug.cgi?id=244789</a>
rdar://99557326

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/Expression.h:
(WGSL::AST::Expression::isArrayAccess const):
Add kind and predicate for ArrayAccess.
* Source/WebGPU/WGSL/AST/Expressions/ArrayAccess.h: Added.
Implementation of AST::ArrayAccess node.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parsePostfixExpression):
Implement parsing of a sequence of StructureAccess or ArrayAccess postfix to an
expression.
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingArrayAccess]):
Test parsing of `x[42]`
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
Add &apos;ArrayAccess.h&apos;
</pre>
----------------------------------------------------------------------
#### 665013d8ad66bf65ac8c8d36c2dd38ba95e21522
<pre>
[WGSL] Implement unary minus expression.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244788">https://bugs.webkit.org/show_bug.cgi?id=244788</a>
rdar://99557228

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/Expression.h:
(WGSL::AST::Expression::isUnaryExpression const):
Add UnaryExpression kind and predicate.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser&lt;Lexer&gt;::parseUnaryExpression):
If a prefix minus token is found, create a UnaryExpression representing the
negation of the expression.
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingUnaryExpression]):
Test parsing of an expression containing a unary minus.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
Add &apos;UnaryExpression.h&apos;
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11290b728b34232a0923b99f927b16e9deb02e08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89546 "check-webkit-style (failure)") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34097 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20258 "Hash 11290b72 for PR 4017 does not build (failure)") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98846 "Built successfully") | 
| | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32591 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28067 "Hash 11290b72 for PR 4017 does not build (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81901 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93256 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95193 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/64/builds/32591 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76415 "Hash 11290b72 for PR 4017 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25828 "Passed tests") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/64/builds/32591 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/43/builds/20258 "Hash 11290b72 for PR 4017 does not build (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68820 "Passed tests") | 
| | [❌ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30348 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/43/builds/20258 "Hash 11290b72 for PR 4017 does not build (failure)") | | 
| | [❌ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30096 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/43/builds/20258 "Hash 11290b72 for PR 4017 does not build (failure)") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33545 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/61/builds/76415 "Hash 11290b72 for PR 4017 does not build (failure)") | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32254 "Hash 11290b72 for PR 4017 does not build (failure)") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/43/builds/20258 "Hash 11290b72 for PR 4017 does not build (failure)") | | 
<!--EWS-Status-Bubble-End-->

djg committed:
[WGSL] Implement parsing for array types and callable array constructors.
https://bugs.webkit.org/show_bug.cgi?id=244792
rdar://99557526

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/TypeDecl.h:
(WGSL::AST::TypeDecl::isArray const):
Add kind and predicate for array types.
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseTypeDecl):
Begin parsing of an array type declaration. Defers to parseArrayType to do
actual parsing.
(WGSL::Parser<Lexer>::parseArrayType):
Parse an array type consisting of 'array' ('<' TypeDecl (','
AdditionExpression)? '>')?
This is different to the WGSL spec as this recursive descent parser can't handle
the grammar factored as AdditiveExpression | BitwiseExpression. Making the
element TypeDecl optional allows the code to parse array in
CallableExpression. Validity of array TypeDecls that require element TypeDecl
will be handled when type-checking the AST.
(WGSL::Parser<Lexer>::parsePrimaryExpression):
Accept array type as a CallableExpression.
* Source/WebGPU/WGSL/ParserPrivate.h:
Add prototype for parsing array type
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
- Test the parsing of a WebGPU example triangle vertex & fragment shader that
contains ArrayType, ArrayAccess and CallableExpression of array.
- Add some macros to help alleviate the repeative AST matching patterns.

litherum commented on Source/WebGPU/WGSL/AST/Expressions/UnaryExpression.h:
```diff
@@ -0,0 +1,57 @@
+namespace WGSL::AST {
+
+enum class UnaryOperation : uint8_t {
+    Negate
```
Worth adding the other kinds? like `&`, `*`, etc?

djg replied:
Following the strategy of implementing what is required for a series of more complicated examples, I'd like to leave each of those for when we need them.

litherum commented on Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
```diff
@@ -209,6 +210,9 @@ - (void)testTrivialGraphicsShader {
 }
 
+#pragma mark -
+#pragma mark Declartions
```
typo

litherum commented on Source/WebGPU/WGSL/Lexer.cpp:
```diff
@@ -236,6 +236,8 @@ Token Lexer<T>::lex()
                 return makeToken(TokenType::KeywordWorkgroup);
             if (view == "write"_s)
                 return makeToken(TokenType::KeywordWrite);
+            if (view == "array"_s)
```
Do we really want to make `array` a keyword? I know the language spec says it is, but why not just treat it as any other type (and say all types can have type parameters, which get matched during type checking)? It would handle `array` and `pointer` systematically instead of as special-cases.

djg replied:
That is a choice and it looks like it was Robin's intention.

litherum commented on Source/WebGPU/WGSL/AST/TypeDecl.h:
```diff
@@ -35,6 +36,7 @@ class TypeDecl : public ASTNode {
 
 public:
     enum class Kind : uint8_t {
+        Array,
```
Arrays are parameterized, are they not? Why not add to the existing `ParameterizedType` class?

djg replied:
I've flip-flopped on this and I bit the bullet and tried to implement array are as `ParameterizedType` but with the possibility of an `Expression` or `TypeDecl` in the 2nd type position, both of which contain `Identifier`, I'm going to go with this code. (eg. `array<f32, foo>` is `foo` a `TypeDecl` or `Expression`?)

I'm happy to re-visit this decision down the road.

litherum commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -329,6 +332,45 @@ Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseTypeDeclAfterIdent
 }
 
+// https://www.w3.org/TR/WGSL/#syntax-array_type_decl
+// ArrayTypeDecl:
```
I was doing this same practice in the WebGPU code, but then people got mad at me so I ended up deleting them. https://github.com/WebKit/WebKit/commit/1c96679bb04a4af96c04cb275c071b2b1c04f9ac

The rationale is that comments should answer "why" not "what". The "what" is already described by the code.

djg replied:
Ok. Personally, I like to have the grammar that it's trying to parse there, so I don't have to go the spec to look it up. But, code comments can get out-of-sync with the ground truth.

litherum replied:
Yeah, spec links are valuable.

litherum commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -329,6 +332,45 @@ Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseTypeDeclAfterIdent
+            // FIXME: According to https://www.w3.org/TR/WGSL/#syntax-element_count_expression
+            // this should be: AdditiveExpression | BitwiseExpression.
+            //
+            // The WGSL grammar doesn't specify expression operator precedence so
```
Link to spec issue?

djg replied:
This is more of an issue with our lack of support for parsing expressions.

litherum commented on Source/WebGPU/WGSL/Parser.cpp:
```diff
@@ -329,6 +332,45 @@ Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseTypeDeclAfterIdent
+// ElementCountExpression:
+//      AdditiveExpression | BitwiseExpression
+template<typename Lexer>
+Expected<UniqueRef<AST::TypeDecl>, Error> Parser<Lexer>::parseArrayType()
```
I guess I already said this in a previous comment, but I'm super surprised we have custom parsing logic for certain types and not others. Seems more elegant to just have a generic parsing function, with full flexibility, and then let type checking figure out what's an error.

I understand that the spec has custom parsing logic for certain types, but does that mean we need to follow the same practice? We only need to behave "as-if" we followed the spec.

litherum commented on Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
```diff
@@ -54,11 +80,11 @@ - (void)testParsingStruct {
-    XCTAssert(shader->structs().size() == 1);
-    XCTAssert(shader->globalVars().isEmpty());
-    XCTAssert(shader->functions().isEmpty());
+    XCTAssertTrue(shader.has_value());
+    XCTAssertTrue(shader->directives().isEmpty());
+    XCTAssertEqual(shader->structs().size(), 1u);
+    XCTAssertTrue(shader->globalVars().isEmpty());
+    XCTAssertTrue(shader->functions().isEmpty());
```
Cool.

litherum commented on Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
```diff
@@ -350,4 +376,100 @@ - (void) testParsingUnaryExpression {
+#pragma mark WebGPU Example Shaders
+
+- (void) testParsingTriangleVert {
+    auto shader = WGSL::parseLChar(
+        "@vertex\n"
+        "fn main(\n"
+        "    @builtin(vertex_index) VertexIndex : u32\n"
+        ") -> @builtin(position) vec4<f32> {\n"
+        "    var pos = array<vec2<f32>, 3>(\n"
+        "        vec2<f32>(0.0, 0.5),\n"
+        "        vec2<f32>(-0.5, -0.5),\n"
+        "        vec2<f32>(0.5, -0.5)\n"
+        "    );\n\n"
+        "    return vec4<f32>(pos[VertexIndex], 0.0, 1.0);\n"
+        "}\n"_s);
```
Awesomeeeeeee

djg committed:
[WGSL] Visitor helper
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/Expressions/ArrayAccess.h:
* Source/WebGPU/WGSL/AST/Expressions/CallableExpression.h:
* Source/WebGPU/WGSL/AST/Expressions/StructureAccess.h:
* Source/WebGPU/WGSL/Forward.h: Copied from Source/WebGPU/WGSL/AST/Expressions/ArrayAccess.h.
* Source/WebGPU/WGSL/Visitor.cpp: Added.
(WGSL::Visitor::hasError const):
(WGSL::Visitor::result):
(WGSL::Visitor::checkErrorAndVisit):
(WGSL::Visitor::maybeCheckErrorAndVisit):
(WGSL::Visitor::visit):
* Source/WebGPU/WGSL/Visitor.h: Added.
(WGSL::Visitor::setError):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

djg committed:
[WGSL] Rename TypeDecl to TypeReference
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

These are not declarations of types but references to types.

* Source/WebGPU/WGSL/AST/Expressions/CallableExpression.h:
* Source/WebGPU/WGSL/AST/FunctionDecl.h:
* Source/WebGPU/WGSL/AST/StructureDecl.h:
* Source/WebGPU/WGSL/AST/TypeReference.h: Renamed from Source/WebGPU/WGSL/AST/TypeDecl.h.
(WGSL::AST::TypeReference::TypeReference):
(WGSL::AST::TypeReference::~TypeReference):
(WGSL::AST::TypeReference::isArray const):
(WGSL::AST::TypeReference::isNamed const):
(WGSL::AST::TypeReference::isParameterized const):
(WGSL::AST::ParameterizedTypeReference::ParameterizedTypeReference):
(WGSL::AST::ParameterizedTypeReference::stringViewToKind):
(WGSL::AST::ParameterizedTypeReference::base const):
(WGSL::AST::ParameterizedTypeReference::elementType):
* Source/WebGPU/WGSL/AST/VariableDecl.h:
* Source/WebGPU/WGSL/Forward.h:
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseStructMember):
(WGSL::Parser<Lexer>::parseTypeReference):
(WGSL::Parser<Lexer>::parseTypeReferenceAfterIdentifier):
(WGSL::Parser<Lexer>::parseArrayType):
(WGSL::Parser<Lexer>::parseVariableDeclWithAttributes):
(WGSL::Parser<Lexer>::parseFunctionDecl):
(WGSL::Parser<Lexer>::parseParameter):
(WGSL::Parser<Lexer>::parsePrimaryExpression):
(WGSL::Parser<Lexer>::parseTypeDecl): Deleted.
(WGSL::Parser<Lexer>::parseTypeDeclAfterIdentifier): Deleted.
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Visitor.cpp:
(WGSL::Visitor::visit):
* Source/WebGPU/WGSL/Visitor.h:
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingStruct]):
(-[WGSLParserTests testParsingGlobalVariable]):
(-[WGSLParserTests testParsingFunctionDecl]):
(-[WGSLParserTests testTrivialGraphicsShader]):
(-[WGSLParserTests testParsingLocalVariable]):
(-[WGSLParserTests testParsingTriangleVert]):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

djg committed:
[WGSL] Decl -> Declaration
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSL/AST/Declaration.h: Renamed from Source/WebGPU/WGSL/AST/Decl.h.
(WGSL::AST::Declaration::Declaration):
(WGSL::AST::Declaration::~Declaration):
(WGSL::AST::Declaration::isVariable const):
(WGSL::AST::Declaration::isStruct const):
(WGSL::AST::Declaration::isFunction const):
* Source/WebGPU/WGSL/AST/Declarations/FunctionDeclaration.h: Renamed from Source/WebGPU/WGSL/AST/FunctionDecl.h.
* Source/WebGPU/WGSL/AST/Declarations/StructDeclaration.h: Renamed from Source/WebGPU/WGSL/AST/StructureDecl.h.
* Source/WebGPU/WGSL/AST/Declarations/TypeDeclaration.h: Added.
(WGSL::AST::TypeReference::TypeReference):
(WGSL::AST::TypeReference::~TypeReference):
(WGSL::AST::TypeReference::isArray const):
(WGSL::AST::TypeReference::isNamed const):
(WGSL::AST::TypeReference::isParameterized const):
(WGSL::AST::ParameterizedTypeReference::ParameterizedTypeReference):
(WGSL::AST::ParameterizedTypeReference::stringViewToKind):
(WGSL::AST::ParameterizedTypeReference::base const):
(WGSL::AST::ParameterizedTypeReference::elementType):
* Source/WebGPU/WGSL/AST/Declarations/VariableDeclaration.h: Renamed from Source/WebGPU/WGSL/AST/VariableDecl.h.
* Source/WebGPU/WGSL/AST/ShaderModule.h:
* Source/WebGPU/WGSL/AST/Statements/VariableStatement.h:
* Source/WebGPU/WGSL/Forward.h:
* Source/WebGPU/WGSL/Parser.cpp:
(WGSL::Parser<Lexer>::parseShader):
(WGSL::Parser<Lexer>::parseGlobalDeclaration):
(WGSL::Parser<Lexer>::parseStructDeclaration):
(WGSL::Parser<Lexer>::parseVariableDeclaration):
(WGSL::Parser<Lexer>::parseVariableDeclarationWithAttributes):
(WGSL::Parser<Lexer>::parseFunctionDeclaration):
(WGSL::Parser<Lexer>::parseStatement):
(WGSL::Parser<Lexer>::parseGlobalDecl): Deleted.
(WGSL::Parser<Lexer>::parseStructDecl): Deleted.
(WGSL::Parser<Lexer>::parseVariableDecl): Deleted.
(WGSL::Parser<Lexer>::parseVariableDeclWithAttributes): Deleted.
(WGSL::Parser<Lexer>::parseFunctionDecl): Deleted.
* Source/WebGPU/WGSL/ParserPrivate.h:
* Source/WebGPU/WGSL/Visitor.cpp:
(WGSL::Visitor::visit):
* Source/WebGPU/WGSL/Visitor.h:
* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingStruct]):
(-[WGSLParserTests testParsingGlobalVariable]):
(-[WGSLParserTests testParsingFunctionDecl]):
(-[WGSLParserTests testTrivialGraphicsShader]):
(-[WGSLParserTests testParsingLocalVariable]):
(-[WGSLParserTests testParsingArrayAccess]):
(-[WGSLParserTests testParsingUnaryExpression]):
(-[WGSLParserTests testParsingTriangleVert]):
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:

djg committed:
[WGSL] Refactor WGSLParserTests
Need the bug URL (OOPS!).
Include a Radar link (OOPS!).

Reviewed by NOBODY (OOPS!).

* Source/WebGPU/WGSLUnitTests/WGSLParserTests.mm:
(-[WGSLParserTests testParsingStruct]):
(-[WGSLParserTests testParsingGlobalVariable]):
(-[WGSLParserTests testParsingFunctionDecl]):
(-[WGSLParserTests testTrivialGraphicsShader]):
(-[WGSLParserTests testParsingLocalVariable]):
(-[WGSLParserTests testParsingArrayAccess]):
(-[WGSLParserTests testParsingUnaryExpression]):
(-[WGSLParserTests testParsingTriangleVert]):
(-[WGSLParserTests testParsingRedFrag]):
